contracts/src/OffChainMessageContracts/ValidatorSetSig.sol

```diff
@@ -0,0 +1,113 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {
+    WarpMessage,
+    IWarpMessenger
+} from "@avalabs/subnet-evm-contracts@1.2.0/contracts/interfaces/IWarpMessenger.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts@4.8.1/security/ReentrancyGuard.sol";
+
+struct ValidatorSetSigMessage {
+    address validatorSetSigAddress;
+    address targetContractAddress;
+    bytes32 targetBlockChainID;
+    uint256 nonce;
+    bytes txPayload;
+}
+
+/**
+ * @dev Contract that verifies that a set threshold of validators from a given blockchainID
+ * have signed an off-chain Warp message and forwards the payload to the destination contract specified in the payload.
+ *
+ * This is intended to be used for safe off-chain governance of enabled contracts by having the target contract be owned by
+ * an instance of this contract and adding the `onlyOwner` modifier to the functions that should be governed.
+ */
+contract ValidatorSetSig is ReentrancyGuard {
+    /**
+     * @notice A blockchain ID whose validators need to sign the message for it to be considered valid.
+     */
+    bytes32 public immutable validatorBlockchainID;
+
+    /**
+     * @dev Tracks nonces for messages sent to a specific contract address to provide replay protection.
+     * The target contract address is used as the key in order to prevent instead of keeping a global nonce in order to prevent
+     * race conditions where the same instance of ValidatorSetSig contract is used to manage multiple downstream contracts.
```

Comment:
I don't think "race condition" is accurate here, since the EVM is a single threaded state machine. IIUC a global nonce would be perfectly safe in terms of replay protection, but would be poor UI since other independent contracts could consume the nonce that the target contract is intending to use.

Reply:
I agree. I wasn't thinking of a software "race condition" here but of human race condition where two parties want to sign different messages at once and one of them ends up going through the process with an already consumed nonce.

Reply:
I just updated the comment, hopefully it's clearer now